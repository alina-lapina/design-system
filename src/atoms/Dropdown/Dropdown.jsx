import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';

const Dropdown = ({
	header, items, onSelect, open, placeholder, searchable, selectedItem,
}) => {
	const id = uuid();
	const node = useRef();
	const [isOpen, setOpen] = useState(open);
	const [availableOptions, filterOptions] = useState(items);
	const [selectedOption, selectItem] = useState(selectedItem || { title: '', id: '' });
	const [inputFieldValue, updateInputValue] = useState('');

	const filterItems = event => {
		updateInputValue(event.target.value);
		filterOptions(items.filter(it => it.title.toLowerCase().includes(event.target.value.toLowerCase())));
	};

	const handleSelection = item => {
		selectItem({ title: item.title, id: item.id });
		onSelect(item);
		setOpen(false);
		filterOptions(items);
		updateInputValue('');
	};

	const handleClickOutside = e => {
		if (!node.current.contains(e.target)) {
			setOpen(false);
		}
	};

	useEffect(() => {
		if (isOpen) {
			document.addEventListener('mousedown', handleClickOutside);
		} else {
			document.removeEventListener('mousedown', handleClickOutside);
		}
		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, [isOpen]);

	return (
		<div className="dropdown-component-wrapper">
			{header && <label htmlFor={id}>{header}</label>}
			<div className="dropdown-interactive-area" ref={node} onClick={() => setOpen(!isOpen)}>
				<input
					className={isOpen ? 'focused' : ''}
					id={id}
					onChange={e => filterItems(e)}
					disabled={!searchable}
					placeholder={selectedOption.title ? selectedOption.title : placeholder}
					value={inputFieldValue}
				/>
				{isOpen && (
					<ul className="list-of-options">
						{availableOptions.map(it => (
							<li
								className={`option-list-element ${selectedOption.id === it.id && 'selected'}`}
								key={it.id}
							>
								<option
									onClick={() => handleSelection(it)}
									value={it.id}
									id={it.id}
								>{it.title}
								</option>
							</li>
						))}
					</ul>
				)}
			</div>
		</div>
	);
};

Dropdown.defaultProps = {
	header: '',
	items: [{ id: '', title: '' }],
	onSelect: () => {},
	open: false,
	searchable: false,
	placeholder: '-- Select --',
};

Dropdown.propTypes = {
	header: PropTypes.string,
	items: PropTypes.arrayOf(PropTypes.shape({
		title: PropTypes.string,
		id: PropTypes.string,
	})),
	onSelect: PropTypes.func,
	open: PropTypes.bool,
	placeholder: PropTypes.string,
	searchable: PropTypes.bool,
	selectedItem: PropTypes.object,
};

export default Dropdown;
