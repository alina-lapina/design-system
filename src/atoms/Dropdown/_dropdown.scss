.dropdown-component-wrapper {
  background: $ssb-white;
  border-radius: 0;
  display: flex;
  flex-direction: column;
  transition: all .25s ease-in-out;

  label {
    margin-bottom: 5px;
  }

  .dropdown-interactive-area {
    cursor: pointer;
    position: relative;

    input {
      border: 1px solid $ssb-dark-5;
      cursor: pointer;
      font-size: 16px;
      height: $input-field-height;
      padding: 4px 34px 4px 10px;

      &::placeholder {
        color: $ssb-dark-6;
      }

      &:hover {
        border: 1px solid $ssb-green-4;
      }

      &:focus, &.focused {
        border: 1px solid $ssb-green-4;
        outline: $ssb-green-4 auto 5px;
        outline-offset: -2px;
      }
    }

    .list-of-options {
      background: $ssb-white;
      border: 1px solid $ice-cold-grey;
      left: 0;
      list-style: none;
      margin: 0;
      max-width: 300px;
      min-width: 150px;
      padding-left: 0;
      position: absolute;
      top: $input-field-height;
      z-index: 9999;

      .option-list-element {
        cursor: pointer;

        option {
          overflow: hidden;
          padding: $global-padding-size;
          text-overflow: ellipsis;
        }

        &:hover {
          background: $ssb-green-2;
        }

        &.selected {
          background: $ssb-green-4;
          color: $ssb-white;
        }
      }
    }

  }
}
