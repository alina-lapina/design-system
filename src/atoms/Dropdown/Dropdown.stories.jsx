import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import Dropdown from './Dropdown';

const items = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	}, {
		title: 'Automobiles',
		id: 'item4',
	},
];

storiesOf('Atoms|Dropdown', module).addDecorator(centered)
	.add('Standard', () => (
		<Dropdown header="Menu header" items={items} />
	))
	.add('Pre-selected value', () => (
		<Dropdown header="Menu header" selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />
	))
	.add('Searchable', () => (
		<Dropdown header="Menu header" searchable items={items} />
	));
