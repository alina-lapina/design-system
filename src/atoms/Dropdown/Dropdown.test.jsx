import React from 'react';
import { shallow } from 'enzyme';
import Dropdown from './Dropdown';

const mockedItems = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	}, {
		title: 'Automobiles',
		id: 'item4',
	},
];

describe('Dropdown component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Dropdown />);
		expect(wrapper).toMatchSnapshot();
	});
	test('Should not be open on init', () => {
		const wrapper = shallow(<Dropdown />);
		expect(wrapper.exists('.list-of-options')).toEqual(false);
	});
	test('List of options should render on click', () => {
		const wrapper = shallow(<Dropdown />);
		wrapper.find('.dropdown-interactive-area').simulate('click');
		expect(wrapper.exists('.list-of-options')).toEqual(true);
	});
	test('Callback should fire on option select', () => {
		const mockedFunction = jest.fn();
		const wrapper = shallow(<Dropdown items={mockedItems} onSelect={e => mockedFunction(e)} />);
		wrapper.find('.dropdown-interactive-area').simulate('click');
		wrapper.find('#item2').simulate('click');
		expect(mockedFunction.mock.calls.length).toBe(1)
	});

	// TODO: Write tests for event listeners
});
