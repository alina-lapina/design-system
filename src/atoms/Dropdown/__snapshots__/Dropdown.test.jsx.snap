// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Matches the snapshot 1`] = `
<div
  className="dropdown-component-wrapper"
>
  <div
    className="dropdown-interactive-area"
    onClick={[Function]}
  >
    <input
      className=""
      disabled={true}
      id="bd30074c-cf84-4d71-b67e-4ab84cdc435d"
      onChange={[Function]}
      placeholder="-- Select --"
      value=""
    />
  </div>
</div>
`;
